lib/dialects/mssql/query-generator.js

```diff
@@ -782,101 +806,40 @@ class MSSQLQueryGenerator extends AbstractQueryGenerator {
     return 'ROLLBACK TRANSACTION;';
   }
 
-  selectFromTableFragment(options, model, attributes, tables, mainTableAs, where) {
-    let topFragment = '';
-    let mainFragment = `SELECT ${attributes.join(', ')} FROM ${tables}`;
-
-    // Handle SQL Server 2008 with TOP instead of LIMIT
-    if (semver.valid(this.sequelize.options.databaseVersion) && semver.lt(this.sequelize.options.databaseVersion, '11.0.0')) {
-      if (options.limit) {
-        topFragment = `TOP ${options.limit} `;
-      }
-      if (options.offset) {
-        const offset = options.offset || 0,
-          isSubQuery = options.hasIncludeWhere || options.hasIncludeRequired || options.hasMultiAssociation;
-        let orders = { mainQueryOrder: [] };
-        if (options.order) {
-          orders = this.getQueryOrders(options, model, isSubQuery);
-        }
-
-        if (!orders.mainQueryOrder.length) {
-          orders.mainQueryOrder.push(this.quoteIdentifier(model.primaryKeyField));
-        }
-
-        const tmpTable = mainTableAs ? mainTableAs : 'OffsetTable';
-        const whereFragment = where ? ` WHERE ${where}` : '';
+  addLimitAndOffset(options) {
+    const composition = new Composition();
+    const isLimit = options.limit !== null && options.limit !== undefined;
 
-        /*
-         * For earlier versions of SQL server, we need to nest several queries
-         * in order to emulate the OFFSET behavior.
-         *
-         * 1. The outermost query selects all items from the inner query block.
-         *    This is due to a limitation in SQL server with the use of computed
-         *    columns (e.g. SELECT ROW_NUMBER()...AS x) in WHERE clauses.
-         * 2. The next query handles the LIMIT and OFFSET behavior by getting
-         *    the TOP N rows of the query where the row number is > OFFSET
-         * 3. The innermost query is the actual set we want information from
-         */
-        const fragment = `SELECT TOP 100 PERCENT ${attributes.join(', ')} FROM ` +
-                        `(SELECT ${topFragment}*` +
-                          ` FROM (SELECT ROW_NUMBER() OVER (ORDER BY ${orders.mainQueryOrder.join(', ')}) as row_num, * ` +
-                            ` FROM ${tables} AS ${tmpTable}${whereFragment})` +
-                          ` AS ${tmpTable} WHERE row_num > ${offset})` +
-                        ` AS ${tmpTable}`;
-        return fragment;
-      }
-      mainFragment = `SELECT ${topFragment}${attributes.join(', ')} FROM ${tables}`;
+    if (options.offset || isLimit) {
+      const offset = options.offset ? new Slot(options.offset) : '0';
+      composition.add('OFFSET ', offset, ' ROWS');
+      if (!isLimit) logger.warn('options.offset was used without options.limit. Most likely this was unintended.');
```

Comment:
Imo this should throw.

Reply:
What I am afraid is breaking people's programs. Perhaps we can issue warnings on the current sequelize version and throw errors on the next one, to developers time to catch up.

Reply:
Isn't this meant to go into v6 anyways?

Reply:
If you don't mind, I prefer to implement this in a separate pull request. I have opened a separate issue #10649 to implement this feature.

Reply:
That's fair.